### DISCUSSION TITLE
Support the -- separator, e.g. `git log -- dir/`

### DISCUSSION THREAD

**Discussion by mytskine** [carapace-sh/carapace-bin#2111] [Q&A] [answered]
### Request

A common pattern is to use `--` to separate flags from other arguments, usually paths. It's a default behaviour for many programs, and libraries like getopt.

### Proposed solution

From the top of my head, some commands that would benefit from this pattern:
- `git status -- some/dir<TAB>` (same with `diff`, `log`, and many other commands)
- `bash -x -- .local/<TAB>`
- `tree -d -- /etc/ng<TAB>`

### Anything else?

I could not find any completion that implements this. I tried to write one myself, but it seems that a `--` argument breaks the parsing of carapace. In my tests, the callback passed to `PositionalAnyCompletion()` was never called for arguments after `--`. Unless I'm mistaken, the carapace API does not document this nor provide an alternative. I did not dig any further, so I may haved missed obvious solutions.

**Comment by rsteube** [comment, marked as answer]
Carapace has support for this with [DashCompletion](https://rsteube.github.io/carapace/carapace/gen/dashCompletion.html) and [DashAnyCompletion](https://rsteube.github.io/carapace/carapace/gen/dashAnyCompletion.html).

I went back and forth to support this implicitly but commands vary too much.
So at least for now it has to be configured explicitly and probably is just not added yet.

To continue the existing positional completion there is [ActionPositional](https://rsteube.github.io/carapace/carapace/defaultActions/actionPositional.html) (ok, this one has no doc yet =D).

```go
carapace.Gen(blameCmd).DashAnyCompletion(
	carapace.ActionPositional(blameCmd),
)
```

> Not entirely happy with this though so it might change a bit in the future.

Git is quite complex so there are also a couple of opiniated choices that were made.
E.g. to separate files from ref arguments (which can be a lot) they are only completed when a path prefix is present (like `./<TAB>`).

**Comment by rsteube** [comment]
Oh and dash completion in carapace resets `Context.Args`.
So `command pos1 pos2 -- pos3 pos4 <TAB>` gets `[pos3, pos4]`.
The complete arguments can still be accessed from the flagset thoug (which is what `ActionPositional` is doing).

The reason for this is that this way `Actions` can be used for dash just as they are for positonal completion.
